### DISCUSSION TITLE
Unable to build ICU-LE-HB

### DISCUSSION THREAD

**Discussion by khushneetinder** [harfbuzz/harfbuzz#3568] [Q&A] [answered]
I am using the ICU libraries for Unicode and Arabic support in my product. To support Silicon architecture on mac, I need to build ICU, Harfbuzz, and Harfbuzz’s layout engine library.  I am able to build ICU and Harfbuzz library as suggested by https://unicode-org.github.io/icu/userguide/layoutengine/paragraph.html.
 
However, I am unable build Harfbuzz’s layout engine library since no build steps are provided at https://github.com/harfbuzz/icu-le-hb . I first tried to run the autogen.sh file and then the errors such as _No package ‘Harfbuzz’ found_ are resolved by adding harfbuzz.pc path to PKF_CONFIG_PATH and similar is done for ICU error.
 
After that, while running make command some compilation errors like headers not found shows up. I didn’t face any error while building ICU and Harfbuzz dylibs.

Please let me know if there are any build steps to build ICU-LE-HB library on Mac and Windows as well.

**Comment by behdad** [comment, marked as answer]
Build instructions are really that of any other package with autotools: run `autogen.sh`, then `make`.  I just build it and works for me. Header-not-found errors are not expected if the pkg-config step was done correctly.

**Reply by khushneetinder** [reply]
I am not sure how can I use the pkg-config. As per steps given in HarfBuzz's Build.md, I have installed the pkg-config using brew.
Can you suggest me how can I use that pkg-config with ICU-LE-HB. I didn't need this while building ICU and Harfbuzz.

**Reply by khaledhosny** [reply]
Can you post the exact commands you ran and the exact error messages you get?

**Reply by khushneetinder** [reply]
### Steps to build ICU, HarfBuzz and ICU-LE-HB 
First of all, I have downloaded the source code of ICU 71.1 from https://github.com/unicode-org/icu/releases/tag/release-71-1. From the source, I have used the _icu4c_ folder and discarded the rest.

From Terminal:
```
cd icu4c/source 
./runConfigureICU MacOSX --disable-layout --disable-layoutex
gnumake CXXFLAGS="-std=c++11"
```


Then I downloaded Harfbuzz source code zip from https://github.com/harfbuzz/harfbuzz
Once downloaded and extracted, I renamed the source code folder _harfbuzz-main_ to _harfbuzz_ and copied it to icu4c/source.

From Terminal:
```
brew install pkg-config ragel gtk-doc freetype glib cairo meson
cd harfbuzz
meson build && meson test -Cbuild
```

Then I download source code zip of ICU-LE-HB from https://github.com/harfbuzz/icu-le-hb
Once downloaded and extracted, I renamed the source code folder _icu-le-hb-master_ to _icu-le-hb_ and copied it to icu4c/source.

From Terminal:
```
cd ../icu-le-hb
sh autogen.sh

```

> Error: *** No autoreconf found, please install it ***

From Terminal
```
brew install automake libtool
sh autogen.sh
```

> Error:
> configure: error: Package requirements (harfbuzz >= 1.2.3) were not met:
> 
> No package 'harfbuzz' found
> 
> Consider adjusting the PKG_CONFIG_PATH environment variable if you
> installed software in a non-standard prefix.
> 
> Alternatively, you may set the environment variables HARFBUZZ_CFLAGS
> and HARFBUZZ_LIBS to avoid the need to call pkg-config.
> See the pkg-config man page for more details.

```
export PKG_CONFIG_PATH=/PATH/TO/icu4c/source/harfbuzz/build/meson-private
make
```

> Error:
> Making all in src
>   CXX      libicu_le_hb_la-LayoutEngine.lo
> LayoutEngine.cpp:18:10: fatal error: 'hb.h' file not found
> #include <hb.h>
>          ^~~~~~
> 1 error generated.
> make[1]: *** [libicu_le_hb_la-LayoutEngine.lo] Error 1
> make: *** [all-recursive] Error 1
> ksingh@192 icu-le-hb % make CXXFLAGS="-std=c++11"
> Making all in src
>   CXX      libicu_le_hb_la-LayoutEngine.lo
>

### Structure:
<img width="1096" alt="structure" src="https://user-images.githubusercontent.com/10447200/167364749-a5a14798-0287-41f7-a79c-10722a365546.png">

**Reply by khaledhosny** [reply]
You built HarfBuzz but didn't install it. If you don't have a particular reason to build HarfBuzz from source, you can install it from Homebrew as well.

**Reply by khushneetinder** [reply]
Thanks for the helping me out. This solution works on Mac Intel machine, but the same issue is not working on M1 machine.
Both machine runs on different OSs. Intel machine has Catalina OS and M1 machine has Monterey OS.

One more thing came into noticed that on M1 machine, brew is installed at /opt/homebrew/bin/brew, whereas on Intel machine, it is installed at /usr/local/bin/brew. Does this has anything to do with following error I am getting while building **icu-le-hb** using _make_ command?

> Making all in src
>   CXX      libicu_le_hb_la-LayoutEngine.lo
> In file included from LayoutEngine.cpp:8:
> ./LETypes.h:17:10: fatal error: 'unicode/uobject.h' file not found
> #include "unicode/uobject.h"
>          ^~~~~~~~~~~~~~~~~~~
> 1 error generated.
> make[1]: *** [libicu_le_hb_la-LayoutEngine.lo] Error 1
> make: *** [all-recursive] Error 1


Does ICU-LE-HB builds on Monterey( M1 machine)?

### Homebrew on Monterey (M1 machine)
![Screenshot 2022-05-09 at 5 51 28 PM](https://user-images.githubusercontent.com/10447200/167408820-563c4592-3a11-475a-a813-bb5e1ec0e03c.png)

**Reply by khaledhosny** [reply]
Did you install ICU on the M1 machine? If you installed it manually, you might want set `PKG_CONFIG_PATH` to where its `.pc` files are installed.

**Reply by khushneetinder** [reply]
Yes, I did. To reverify it I ran the command `brew reinstall icu4c` and
`export PKG_CONFIG_PATH=$PKG_CONFIG_PATH:/PATH/TO/icu4c/source/config` prior to make command.

However, on running `sh autogen.sh`, I am getting the following error.

> configure: error: Package requirements (icu-uc) were not met:
> 
> No package 'icu-uc' found
> 
> Consider adjusting the PKG_CONFIG_PATH environment variable if you
> installed software in a non-standard prefix.
> 
> Alternatively, you may set the environment variables ICU_CFLAGS
> and ICU_LIBS to avoid the need to call pkg-config.
> See the pkg-config man page for more details.

It seems like there is some other library which needs to be installed using Homebrew. What steps I need to perform so that ICU can be installed properly since `brew install icu4c` doesn't work alone.

**Reply by khaledhosny** [reply]
On Homebrew, `icu4c` is a ”keg-only” package, here is what `brew info icu4c` says:
```
==> Caveats
icu4c is keg-only, which means it was not symlinked into /opt/homebrew,
because macOS provides libicucore.dylib (but nothing else).

If you need to have icu4c first in your PATH, run:
  echo 'export PATH="/opt/homebrew/opt/icu4c/bin:$PATH"' >> /Users/khaled/.bash_profile
  echo 'export PATH="/opt/homebrew/opt/icu4c/sbin:$PATH"' >> /Users/khaled/.bash_profile

For compilers to find icu4c you may need to set:
  export LDFLAGS="-L/opt/homebrew/opt/icu4c/lib"
  export CPPFLAGS="-I/opt/homebrew/opt/icu4c/include"

For pkg-config to find icu4c you may need to set:
  export PKG_CONFIG_PATH="/opt/homebrew/opt/icu4c/lib/pkgconfig"
```

**Comment by khushneetinder** [comment]
After the steps suggested by you, I am able to build ICU-LE-HB.

Next, I am following these steps for building the Paragraph Layout library with HarfBuzz as mention on https://unicode-org.github.io/icu/userguide/layoutengine/paragraph.html

> While the ICU LayoutEngine is deprecated as of ICU 54, the ICU Paragraph Layout library is not. The Paragraph Layout library must now be built using the HarfBuzz engine instead of the ICU LayoutEngine.
> 
> UNIX MAKEFILE INSTRUCTIONS / CYGWIN / MSYS / ETC. (ICU 54+)
> The following steps must be completed in order:
> 
> 1. Build and install a complete ICU with the --disable-layout --disable-layoutex switches passed to configure
> 1. Build and install HarfBuzz - http://harfbuzz.org (HarfBuzz’s use of ICU may be enabled or disabled at your choice)
> 1. Build and install the [icu-le-hb](http://harfbuzz.org/) library.
> 1. Now, rerun “configure” on the exact same ICU workspace used above:
>     1.  with “icu-le-hb” AND the above-mentioned installed ICU available via pkg-config ( pkg-config --modversion icu-le-hb should return a version, such as “0.0.0” )
>    1. with the –disable-layout --enable-layoutex switches passed to configure
> 1. next, run make install JUST in the source/layoutex directory, to install libiculx and icu-lx.pc

As mentioned in steps I tried to install the icu-le-hb library using `brew install icu-le-hb` but I got following error

> Error:
> Warning: No available formula with the name "icu-le-hb".
> ==> Searching for similarly named formulae...
> Error: No similarly named formulae found.
> ==> Searching for a previously deleted formula (in the last month)...
> Error: No previously deleted formula found.
> ==> Searching taps on GitHub...
> Error: No formulae found in taps.
> 


Then I try to ignore the error just to check whether next steps 4 and 5 respectively would run or not with following commands, but I again got an error.

From Terminal 
```
cd ..
./runConfigureICU MacOSX --disable-layout --enable-layoutex
cd layoutex
make install
```

> Error:
> /bin/sh ../mkinstalldirs /usr/local/include/layout
> mkdir /usr/local/include/layout
> mkdir: /usr/local/include/layout: Permission denied
> make: *** [install-headers] Error 1

Note:

> `pkg-config --modversion icu-le-hb` returns _1.0.3_ after 
> `export PKG_CONFIG_PATH=$PKG_CONFIG_PATH:/PATH/TO/icu4c/source/icu-le-hb`

From Terminal 
`sudo make install`

> Error:
> objc[8617]: Class AMSupportURLConnectionDelegate is implemented in both /usr/lib/libamsupport.dylib (0x1fcef7130) and /Library/Apple/System/Library/PrivateFrameworks/MobileDevice.framework/Versions/A/MobileDevice (0x1042302c8). One of the two will be used. Which one is undefined.
> objc[8617]: Class AMSupportURLSession is implemented in both /usr/lib/libamsupport.dylib (0x1fcef7180) and /Library/Apple/System/Library/PrivateFrameworks/MobileDevice.framework/Versions/A/MobileDevice (0x104230318). One of the two will be used. Which one is undefined.
> /bin/sh ../mkinstalldirs /usr/local/include/layout
> mkdir /usr/local/include/layout
> /usr/bin/install -c -m 644 ./layout/ParagraphLayout.h /usr/local/include/layout
> /usr/bin/install -c -m 644 ./layout/RunArrays.h /usr/local/include/layout
> /usr/bin/install -c -m 644 ./layout/playout.h /usr/local/include/layout
> /usr/bin/install -c -m 644 ./layout/plruns.h /usr/local/include/layout
>    clang++	 ...  LXUtilities.cpp
> objc[8634]: Class AMSupportURLConnectionDelegate is implemented in both /usr/lib/libamsupport.dylib (0x1fcef7130) and /Library/Apple/System/Library/PrivateFrameworks/MobileDevice.framework/Versions/A/MobileDevice (0x107fbc2c8). One of the two will be used. Which one is undefined.
> objc[8634]: Class AMSupportURLSession is implemented in both /usr/lib/libamsupport.dylib (0x1fcef7180) and /Library/Apple/System/Library/PrivateFrameworks/MobileDevice.framework/Versions/A/MobileDevice (0x107fbc318). One of the two will be used. Which one is undefined.
> LXUtilities.cpp:10:10: fatal error: 'layout/LETypes.h' file not found
> #include "layout/LETypes.h"
>          ^~~~~~~~~~~~~~~~~~
> 1 error generated.
> *** Failed compilation command follows: ----------------------------------------------------------
> clang++ -DU_ATTRIBUTE_DEPRECATED= -DU_LAYOUTEX_IMPLEMENTATION -I/opt/homebrew/opt/icu4c/include -DU_HAVE_STRTOD_L=1 -DU_HAVE_XLOCALE_H=1 -DU_HAVE_STRING_VIEW=1 -I. -I./unicode -I../common -I./.. -O2 -W -Wall -pedantic -Wpointer-arith -Wwrite-strings -Wno-long-long -std=c++11 -Wno-ambiguous-reversed-operator -Qunused-arguments -Wno-parentheses-equality -Wglobal-constructors -fvisibility=hidden -fno-common -c -dynamic -MMD -MT LXUtilities.d LXUtilities.o LXUtilities.ao -o LXUtilities.o LXUtilities.cpp
> --- ( rebuild with "/Applications/Xcode.app/Contents/Developer/usr/bin/make VERBOSE=1 install" to show all parameters ) --------
> make: *** [LXUtilities.o] Error 1

From Terminal
```
export ICULEHB_CFLAGS="-I/usr/local/include/icu-le-hb"       
export ICULEHB_LIBS="-L/usr/local/lib -licu-le-hb"
sudo make install
```

> Error:
> /bin/sh ../mkinstalldirs /usr/local/include/layout
> /usr/bin/install -c -m 644 ./layout/ParagraphLayout.h /usr/local/include/layout
> /usr/bin/install -c -m 644 ./layout/RunArrays.h /usr/local/include/layout
> /usr/bin/install -c -m 644 ./layout/playout.h /usr/local/include/layout
> /usr/bin/install -c -m 644 ./layout/plruns.h /usr/local/include/layout
>    clang++	 ...  LXUtilities.cpp
> LXUtilities.cpp:10:10: fatal error: 'layout/LETypes.h' file not found
> #include "layout/LETypes.h"
>          ^~~~~~~~~~~~~~~~~~
> 1 error generated.
> *** Failed compilation command follows: ----------------------------------------------------------
> clang++ -DU_ATTRIBUTE_DEPRECATED= -DU_LAYOUTEX_IMPLEMENTATION -I/opt/homebrew/opt/icu4c/include -DU_HAVE_STRTOD_L=1 -DU_HAVE_XLOCALE_H=1 -DU_HAVE_STRING_VIEW=1 -I. -I./unicode -I../common -I./.. -O2 -W -Wall -pedantic -Wpointer-arith -Wwrite-strings -Wno-long-long -std=c++11 -Wno-ambiguous-reversed-operator -Qunused-arguments -Wno-parentheses-equality -Wglobal-constructors -fvisibility=hidden -fno-common -c -dynamic -MMD -MT LXUtilities.d LXUtilities.o LXUtilities.ao -o LXUtilities.o LXUtilities.cpp
> --- ( rebuild with "/Applications/Xcode.app/Contents/Developer/usr/bin/make VERBOSE=1 install" to show all parameters ) --------
> make: *** [LXUtilities.o] Error 1

What remaining steps should I follow so that I can able to build and install all three libraries?

**Reply by khushneetinder** [reply]
A one more thing comes into noticed that library linked path of Harfbuzz, ICU, and ICU-LE-HB is different than the ICU 61. For instance , the `otool -l` command for ICU 61 returns 

> 	@executable_path/../Frameworks/libharfbuzz.dylib (compatibility version 20101.0.0, current version 20101.0.0)
> 	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1252.50.4)

whereas ICU 71.1 returns 

> /usr/local/lib/libicu-le-hb.0.dylib (compatibility version 1.0.0, current version 1.0.0)
> 	/opt/homebrew/opt/harfbuzz/lib/libharfbuzz.0.dylib (compatibility version 40201.0.0, current version 40201.0.0)
> 	/opt/homebrew/opt/icu4c/lib/libicuuc.70.dylib (compatibility version 70.0.0, current version 70.1.0)
> 	/opt/homebrew/opt/icu4c/lib/libicudata.70.dylib (compatibility version 70.0.0, current version 70.1.0)
> 	/usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 1200.3.0)
> 	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1311.0.0)

Please let me know is it expected or I am doing it wrong. Is it related to the Incomplete 5th step?

**Comment by khushneetinder** [comment]
@khaledhosny Could you please help us out to build ICU. I have tried to solve the 5th step in multiple ways, but didn't get succeeded.

**Reply by khaledhosny** [reply]
Sorry, but I can't provide much help. Mixing self compiled and brew installed packages can be tricky, and I can't provide much help here.

**Reply by srl295** [reply]
Hi @khushneetinder — Came here because of your icu4c post. (hullo @behdad and @khaledhosny ! ) I note that:
- icu-le-hb has been updated since you posted this, it is now on v 1.2.3

> `pkg-config --modversion icu-le-hb` returns 1.0.3 after
> `export PKG_CONFIG_PATH=$PKG_CONFIG_PATH:/PATH/TO/icu4c/source/icu-le-hb`

this `export` is required.  You should NOT set `ICULEHB_CFLAGS` or `ICULEHB_LIBS`. 

Did you re-run `./runConfigureICU MacOSX --disable-layout --enable-layoutex` AFTER setting PKG_CONFIG_PATH properly?

It should detect icu-le-hb as part of that configure step.
